code/menuui/readyroom.cpp

```diff
@@ -1548,20 +1548,29 @@ void campaign_room_scroll_info_down()
 		gamesnd_play_iface(InterfaceSounds::GENERAL_FAIL);
 }
 
-void campaign_reset(const SCP_string& campaign_file) {
+void campaign_reset(const SCP_string& campaign_file)
+{
 	auto filename = campaign_file + FS_CAMPAIGN_FILE_EXT;
 
 	mission_campaign_savefile_delete(filename.c_str());
-	mission_campaign_load(filename.c_str(), nullptr, 1 , false); // retail doesn't reset stats when resetting the campaign
-	mission_campaign_next_mission();
 
-	// Goober5000 - reinitialize tech database if needed
-	if ((Campaign.flags & CF_CUSTOM_TECH_DATABASE) || !stricmp(Campaign.filename, "freespace2")) {
-		// reset tech database to what's in the tables
-		tech_reset_to_default();
+	int load_status = mission_campaign_load(filename.c_str(), nullptr, 1 , false); // retail doesn't reset stats when resetting the campaign
+
+	// see if we successfully loaded this campaign
+	if (load_status == 0) {
+		// Goober5000 - reinitialize tech database if needed
+		if ((Campaign.flags & CF_CUSTOM_TECH_DATABASE) || !stricmp(Campaign.filename, "freespace2")) {
+			// reset tech database to what's in the tables
+			tech_reset_to_default();
+
+			// write the savefile so that we don't later load a stale techroom
+			Pilot.save_savefile();
```

Comment:
I don't understand the question.  The player never chooses to save anything; it's always automatic.  In this particular case, the fact that the tech database has been reset must be saved, because it is campaign-related data that has been modified, and it is reloaded from the file a few screens later.

Reply:
The only cases I can think of where the player can avoid saving/choose not to save:
1. During a mission, when they hit Esc and select "Yes, Close"
2. After a mission, when they click on Retry.

But neither of those can happen here, right?

Reply:
Correct.  This save logic is completely independent of the mission save logic.  Furthermore, the mission save logic already has code to back out any changes in the tech db if the player chooses not to accept.